File: func_parser/parser/tokenizer.py
"""Tokenizer for the func_parser command language."""
from __future__ import annotations

import enum
import re
from dataclasses import dataclass
from typing import List

__all__ = ["TokenType", "Token", "Tokenizer"]


class TokenType(enum.Enum):
    """All token types produced by the tokenizer."""
    COMMAND = "command"
    ARG = "arg"
    PIPE = "pipe"
    AND = "and"
    OR = "or"
    REDIRECT_OUT = "redirect_out"
    REDIRECT_APPEND = "redirect_append"
    REDIRECT_CLIPBOARD = "redirect_clipboard"
    SET_VAR = "set_var"
    EXECUTE = "execute"
    COMMENT = "comment"
    EOF = "eof"


@dataclass
class Token:
    """A single token produced by the tokenizer."""
    type: TokenType
    value: str
    pos: int


class Tokenizer:
    """Converts a raw input string into a list of :class:`Token` objects.

    Handles:
    - Quoted strings (``"..."`` and ``'...'``)
    - Escape sequences inside quotes
    - ``{file.txt}`` argument injection markers
    - ``${var}`` / ``$VAR`` substitution markers (preserved as-is)
    - Operators: ``|``, ``&&``, ``||``, ``>``, ``>>``, ``>clipboard``
    - Special commands: ``//set``, ``//setenv``, ``/execute``
    - Comments (``#``)
    """

    def tokenize(self, text: str) -> List[Token]:
        """Parse *text* and return a list of tokens."""
        tokens: List[Token] = []
        i = 0
        length = len(text)

        while i < length:
            # Skip leading whitespace
            if text[i].isspace():
                i += 1
                continue

            # Comment
            if text[i] == "#":
                tokens.append(Token(TokenType.COMMENT, text[i:], i))
                break

            # >>clipboard / >> (must check >> before >)
            if text[i:i+2] == ">>":
                rest = text[i+2:].lstrip()
                if rest.startswith("clipboard"):
                    tokens.append(Token(TokenType.REDIRECT_CLIPBOARD, ">>clipboard", i))
                    i += 2 + len(text[i+2:]) - len(rest) + len("clipboard")
                else:
                    tokens.append(Token(TokenType.REDIRECT_APPEND, ">>", i))
                    i += 2
                continue

            # >clipboard / >
            if text[i] == ">" and text[i:i+2] != ">>":
                rest = text[i+1:].lstrip()
                if rest.startswith("clipboard"):
                    tokens.append(Token(TokenType.REDIRECT_CLIPBOARD, ">clipboard", i))
                    i += 1 + len(text[i+1:]) - len(rest) + len("clipboard")
                else:
                    tokens.append(Token(TokenType.REDIRECT_OUT, ">", i))
                    i += 1
                continue

            # || (before |)
            if text[i:i+2] == "||":
                tokens.append(Token(TokenType.OR, "||", i))
                i += 2
                continue

            # &&
            if text[i:i+2] == "&&":
                tokens.append(Token(TokenType.AND, "&&", i))
                i += 2
                continue

            # |
            if text[i] == "|":
                tokens.append(Token(TokenType.PIPE, "|", i))
                i += 1
                continue

            # Quoted string
            if text[i] in ('"', "'"):
                value, new_i = self._read_quoted(text, i)
                tokens.append(Token(TokenType.ARG, value, i))
                i = new_i
                continue

            # {file.txt} injection marker
            if text[i] == "{":
                end = text.find("}", i)
                if end != -1:
                    value = text[i:end+1]
                    tokens.append(Token(TokenType.ARG, value, i))
                    i = end + 1
                    continue

            # Word token (command, arg, set, execute, …)
            word, new_i = self._read_word(text, i)
            token_type = self._classify_word(word, tokens)
            tokens.append(Token(token_type, word, i))
            i = new_i

        tokens.append(Token(TokenType.EOF, "", length))
        return tokens

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _read_quoted(self, text: str, start: int) -> tuple[str, int]:
        """Read a quoted string starting at *start*.  Returns (value, next_index)."""
        quote = text[start]
        i = start + 1
        buf: list[str] = []
        while i < len(text):
            ch = text[i]
            if ch == "\\" and i + 1 < len(text):
                next_ch = text[i + 1]
                escape_map = {"n": "\n", "t": "\t", "r": "\r", "\\": "\\", quote: quote}
                buf.append(escape_map.get(next_ch, next_ch))
                i += 2
            elif ch == quote:
                i += 1
                break
            else:
                buf.append(ch)
                i += 1
        return "".join(buf), i

    def _read_word(self, text: str, start: int) -> tuple[str, int]:
        """Read a non-quoted word (stops at whitespace or operators)."""
        i = start
        while i < len(text):
            ch = text[i]
            if ch.isspace():
                break
            if ch in ("|", "&", ">", "#"):
                # Don't consume operator characters as part of a word
                # unless they're part of a //-prefixed directive
                if i == start:
                    i += 1  # consume at least one char to avoid infinite loop
                break
            i += 1
        return text[start:i], i

    def _classify_word(self, word: str, preceding: List[Token]) -> TokenType:
        """Determine the token type for *word* given what has been seen so far."""
        # Special directives
        low = word.lower()
        if low in ("//set", "//setenv"):
            return TokenType.SET_VAR
        if low == "/execute":
            return TokenType.EXECUTE

        # First real token on line (or after operator) → COMMAND if starts with /
        meaningful = [t for t in preceding if t.type not in (TokenType.EOF, TokenType.COMMENT)]
        if not meaningful:
            return TokenType.COMMAND

        last = meaningful[-1]
        if last.type in (TokenType.PIPE, TokenType.AND, TokenType.OR):
            return TokenType.COMMAND

        return TokenType.ARG

File: func_parser_rs/src/tokenizer.rs
//! Tokenizer for func_parser_rs.

/// All token types.
#[derive(Debug, Clone, PartialEq)]
pub enum TokenType {
    Command,
    Arg,
    Pipe,
    And,
    Or,
    RedirectOut,
    RedirectAppend,
    RedirectClipboard,
    SetVar,
    Execute,
    Comment,
    Eof,
}

/// A single token.
#[derive(Debug, Clone)]
pub struct Token {
    pub kind: TokenType,
    pub value: String,
    pub pos: usize,
}

impl Token {
    pub fn new(kind: TokenType, value: impl Into<String>, pos: usize) -> Self {
        Self { kind, value: value.into(), pos }
    }
}

/// Converts a raw input string into a list of `Token` objects.
pub struct Tokenizer;

impl Tokenizer {
    pub fn new() -> Self {
        Self
    }

    pub fn tokenize(&self, text: &str) -> Vec<Token> {
        let mut tokens = Vec::new();
        let chars: Vec<char> = text.chars().collect();
        let length = chars.len();
        let mut i = 0;

        while i < length {
            // Skip whitespace
            if chars[i].is_whitespace() {
                i += 1;
                continue;
            }

            // Comment
            if chars[i] == '#' {
                tokens.push(Token::new(TokenType::Comment, &text[i..], i));
                break;
            }

            // Check for >> (before >)
            if i + 1 < length && chars[i] == '>' && chars[i + 1] == '>' {
                // Look for "clipboard" after >>
                let rest_start = i + 2;
                let rest = text[rest_start..].trim_start();
                if rest.starts_with("clipboard") {
                    tokens.push(Token::new(TokenType::RedirectClipboard, ">>clipboard", i));
                    i = rest_start + (text[rest_start..].len() - rest.len()) + "clipboard".len();
                } else {
                    tokens.push(Token::new(TokenType::RedirectAppend, ">>", i));
                    i += 2;
                }
                continue;
            }

            // Check for >
            if chars[i] == '>' {
                let rest_start = i + 1;
                let rest = text[rest_start..].trim_start();
                if rest.starts_with("clipboard") {
                    tokens.push(Token::new(TokenType::RedirectClipboard, ">clipboard", i));
                    i = rest_start + (text[rest_start..].len() - rest.len()) + "clipboard".len();
                } else {
                    tokens.push(Token::new(TokenType::RedirectOut, ">", i));
                    i += 1;
                }
                continue;
            }

            // || (before |)
            if i + 1 < length && chars[i] == '|' && chars[i + 1] == '|' {
                tokens.push(Token::new(TokenType::Or, "||", i));
                i += 2;
                continue;
            }

            // &&
            if i + 1 < length && chars[i] == '&' && chars[i + 1] == '&' {
                tokens.push(Token::new(TokenType::And, "&&", i));
                i += 2;
                continue;
            }

            // |
            if chars[i] == '|' {
                tokens.push(Token::new(TokenType::Pipe, "|", i));
                i += 1;
                continue;
            }

            // Quoted string
            if chars[i] == '"' || chars[i] == '\'' {
                let (value, new_i) = self.read_quoted(&chars, i);
                tokens.push(Token::new(TokenType::Arg, value, i));
                i = new_i;
                continue;
            }

            // {file.txt} injection
            if chars[i] == '{' {
                if let Some(end) = text[i..].find('}') {
                    let value = &text[i..i + end + 1];
                    tokens.push(Token::new(TokenType::Arg, value, i));
                    i += end + 1;
                    continue;
                }
            }

            // Word token
            let (word, new_i) = self.read_word(&chars, i);
            let kind = self.classify_word(&word, &tokens);
            tokens.push(Token::new(kind, word, i));
            i = new_i;
        }

        tokens.push(Token::new(TokenType::Eof, "", length));
        tokens
    }

    fn read_quoted(&self, chars: &[char], start: usize) -> (String, usize) {
        let quote = chars[start];
        let mut i = start + 1;
        let mut buf = String::new();
        while i < chars.len() {
            let ch = chars[i];
            if ch == '\\' && i + 1 < chars.len() {
                let next = chars[i + 1];
                let escaped = match next {
                    'n' => '\n',
                    't' => '\t',
                    'r' => '\r',
                    '\\' => '\\',
                    c if c == quote => quote,
                    c => c,
                };
                buf.push(escaped);
                i += 2;
            } else if ch == quote {
                i += 1;
                break;
            } else {
                buf.push(ch);
                i += 1;
            }
        }
        (buf, i)
    }

    fn read_word(&self, chars: &[char], start: usize) -> (String, usize) {
        let mut i = start;
        while i < chars.len() {
            let ch = chars[i];
            if ch.is_whitespace() {
                break;
            }
            if matches!(ch, '|' | '&' | '>' | '#') {
                if i == start {
                    i += 1;
                }
                break;
            }
            i += 1;
        }
        let word: String = chars[start..i].iter().collect();
        (word, i)
    }

    fn classify_word(&self, word: &str, preceding: &[Token]) -> TokenType {
        let low = word.to_lowercase();
        if low == "//set" || low == "//setenv" {
            return TokenType::SetVar;
        }
        if low == "/execute" {
            return TokenType::Execute;
        }
        let meaningful: Vec<&Token> = preceding
            .iter()
            .filter(|t| !matches!(t.kind, TokenType::Eof | TokenType::Comment))
            .collect();

        if meaningful.is_empty() {
            // Only classify as Command if the word starts with '/'
            if word.starts_with('/') {
                return TokenType::Command;
            }
            return TokenType::Arg;
        }

        if let Some(last) = meaningful.last() {
            if matches!(last.kind, TokenType::Pipe | TokenType::And | TokenType::Or) {
                return TokenType::Command;
            }
        }
        TokenType::Arg
    }
}

impl Default for Tokenizer {
    fn default() -> Self {
        Self::new()
    }
}
